### PR TITLE
Add pallet-invulnerables (#163)

### PR DESCRIPTION
**Pull request by rimbi**
This pallet mimics what we do in collator-selection:
- A storage item that holds the accounts that are invulnerables
- A extrinsic that updates this storage item via root origin

We would also need to include a migration to migrate those invulnerables from pallet-collator-selection to pallet-invulnerables.

### CODE REVIEW

**Comment by girazoki** [comment]
@rimbi Can you provide a description of what the PR does?

File: pallets/invulnerables/src/migration.rs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+/// Version 1 Migration
+/// This migration ensures that any existing `Invulnerables` storage lists are sorted.
+pub mod v1 {
+    use super::*;
+    use frame_support::pallet_prelude::*;
+    #[cfg(feature = "try-runtime")]
+    use sp_std::prelude::*;
+
+    pub struct MigrateToV1<T>(sp_std::marker::PhantomData<T>);
+    impl<T: Config> OnRuntimeUpgrade for MigrateToV1<T> {
```
Comment (girazoki) [inline pallets/invulnerables/src/migration.rs:31]: I dont understand this migraiton. At the beginning there wont be any invulnerables. What we need to do is set them for the first time, but that has nothing to do with the storage version of the pallet (as it did not change). You can remove this fiole for now, as we need to do the migration in some other way

Comment (rimbi) [reply pallets/invulnerables/src/migration.rs:31]: I assumed we would work on it together. You can ignore it for now.

**Comment by girazoki** [comment]
At some point I think it might be good to offer this pallet-abstraction to cumulus, leaving this comment here to remember :)

File: runtime/dancebox/src/migrations.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        assert_eq!(
+            number_of_invulnerables, stored_invulnerables_len,
+            "after migration, there should be the same number of invulnerables"
+        );
+
+        Ok(())
+    }
+}
+
+pub struct CommonMigrations<Runtime>(PhantomData<Runtime>);
```
Comment (girazoki) [inline runtime/dancebox/src/migrations.rs:84]:
```suggestion
pub struct DanceboxMigrations<Runtime>(PhantomData<Runtime>);
```

File: runtime/dancebox/src/lib.rs [hunk 3]
```diff
@@ -669,7 +679,7 @@ impl pallet_proxy::Config for Runtime {
 
 impl pallet_migrations::Config for Runtime {
     type RuntimeEvent = RuntimeEvent;
-    type MigrationsList = ();
+    type MigrationsList = (migrations::CommonMigrations<Runtime>,);
```
Comment (girazoki) [inline runtime/dancebox/src/lib.rs:682]:
```suggestion
    type MigrationsList = (migrations::DanceboxMigrations<Runtime>,);
```

File: runtime/dancebox/src/migrations.rs (hunk 2 above)
Comment (rimbi) [reply runtime/dancebox/src/migrations.rs]: done

File: runtime/dancebox/src/lib.rs (hunk 3 above)
Comment (rimbi) [reply runtime/dancebox/src/lib.rs]: done

File: runtime/dancebox/tests/integration_test.rs [hunk 4]
```diff
@@ -1983,3 +1986,34 @@ fn check_well_known_keys() {
...
+        ])
+        .with_config(pallet_configuration::HostConfiguration {
+            max_collators: 100,
+            min_orchestrator_collators: 2,
+            max_orchestrator_collators: 2,
+            collators_per_container: 2,
+        })
+        .build()
+        .execute_with(|| {
+            let invulnerables_collator_selection = CollatorSelection::invulnerables();
```
Comment (girazoki) [inline runtime/dancebox/tests/integration_test.rs:2010]: One suggestion now that we got this test working: remove the

```
.with_collators(vec![
            (AccountId::from(ALICE), 210 * UNIT),
            (AccountId::from(BOB), 100 * UNIT),
        ])
```

and use `put_storage_value` like in https://github.com/moonbeam-foundation/moonbeam/blob/446faeb04153d127211a499244c900a1a5c71e54/pallets/asset-manager/src/tests.rs#L468. That way we are ready to remove the CollatorSelection in the following PR without having to change the test

**Review by girazoki — APPROVED** [review]
A small comment regarding the migration test, but otherwise looking good!

File: runtime/dancebox/tests/integration_test.rs (hunk 4 above)
Comment (rimbi) [reply runtime/dancebox/tests/integration_test.rs]: done

**Pull request merged**